From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Michael Ludwig <michaelludwig@google.com>
Date: Fri, 25 Sep 2026 10:47:17 -0700
Subject: [M152] [skia] Update Ganesh flush calls to propagate success

Original change's description:
> [skia] Update Ganesh flush calls to propagate success
>
> Ganesh's flush now reports success, so update call sites to include that
> in their `success` results that were already being propagated up.
> Additionaly, raster_decoder's SetCleared() logic is unified now between
> Ganesh and Graphite to only mark the shared image as cleared after the
> flush is successful.
>
> Bug: 561994362
> Fixed: 561994362
> Change-Id: I8dbe91d8f6481af4542ff8f4023811efa6966f87
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8449772
> Commit-Queue: Michael Ludwig <michaelludwig@google.com>
> Reviewed-by: Kyle Charbonneau <kylechar@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1703817}

(cherry picked from commit de4926eaa4d86714638bb61ee68221a03c64ffbd)

Bug: 566046238,561994362,561994362
Change-Id: I8dbe91d8f6481af4542ff8f4023811efa6966f87
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8467690
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Commit-Queue: Michael Ludwig <michaelludwig@google.com>
Auto-Submit: chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Cr-Commit-Position: refs/branch-heads/7977@{#2733}
Cr-Branched-From: b7fe14017379ddffae396d944fb8b59a5896c261-refs/heads/main@{#1669021}

diff --git a/gpu/command_buffer/service/raster_decoder.cc b/gpu/command_buffer/service/raster_decoder.cc
index 5658154ace175e06ab89e1362db3afaf05353c47..b1a796c62588bc9d2bcda7aa6c5ca4867a3a0d07 100644
--- a/gpu/command_buffer/service/raster_decoder.cc
+++ b/gpu/command_buffer/service/raster_decoder.cc
@@ -2969,13 +2969,10 @@ void RasterDecoderImpl::DoBeginRasterCHROMIUM(GLfloat r,
   // incorrect.
   if (needs_clear) {
     raster_canvas_->drawColor(sk_color_4f, SkBlendMode::kSrc);
-    if (graphite_shared_context()) {
-      should_clear_shared_image_ = true;
-    } else {
-      shared_image_->SetCleared();
-    }
+    // Remember that we can mark the shared image as cleared if the deferred
+    // GPU work is flushed successfully.
+    should_clear_shared_image_ = true;
   }
-  DCHECK(graphite_shared_context() || shared_image_->IsCleared());
 }
 
 scoped_refptr<Buffer> RasterDecoderImpl::GetShmBuffer(uint32_t shm_id) {
diff --git a/gpu/command_buffer/service/shared_context_state.cc b/gpu/command_buffer/service/shared_context_state.cc
index 89eb25cc6ba1f6b800caab2e48f20d6eec20f0fa..11edb79c32b45b216089d86bc2907dd2eb5ebc93 100644
--- a/gpu/command_buffer/service/shared_context_state.cc
+++ b/gpu/command_buffer/service/shared_context_state.cc
@@ -844,8 +844,9 @@ bool SharedContextState::FlushAndSubmit(bool sync_to_cpu) {
                                           : skgpu::graphite::SyncToCpu::kNo);
     return flush_succeeded;
   } else if (gr_context()) {
-    gr_context()->flush();
-    return gr_context()->submit(sync_to_cpu ? GrSyncCpu::kYes : GrSyncCpu::kNo);
+    return gr_context()
+        ->flushAndSubmit(sync_to_cpu ? GrSyncCpu::kYes : GrSyncCpu::kNo)
+        .fSuccess;
   }
   return true;
 }
@@ -871,7 +872,7 @@ bool SharedContextState::FlushWriteAccess(
       for (int plane_index = 0; plane_index < num_planes; plane_index++) {
         auto* surface = access->surface(plane_index);
         DCHECK(surface);
-        skgpu::ganesh::Flush(surface);
+        success &= skgpu::ganesh::Flush(surface).fSuccess;
       }
     }
   }
diff --git a/gpu/command_buffer/service/webgpu_decoder_impl.cc b/gpu/command_buffer/service/webgpu_decoder_impl.cc
index 74ed3a371cc53488f80c15e09eacf988c59e98ad..57a3b79bc721b29b1fc8fbe6b41fa3cf2c49b032 100644
--- a/gpu/command_buffer/service/webgpu_decoder_impl.cc
+++ b/gpu/command_buffer/service/webgpu_decoder_impl.cc
@@ -855,7 +855,7 @@ class WebGPUDecoderImpl final : public WebGPUDecoder {
       // will populate it with semaphores and call GrDirectContext::flush.
       success = true;
       if (shared_context_state_->gr_context()) {
-        skgpu::ganesh::Flush(surface);
+        success = skgpu::ganesh::Flush(surface).fSuccess;
       } else {
         DCHECK(shared_context_state_->graphite_shared_context());
         DCHECK(shared_context_state_->gpu_main_graphite_recorder());
@@ -2469,7 +2469,10 @@ bool WebGPUDecoderImpl::ClearSharedImageWithSkia(const Mailbox& mailbox) {
   // It's ok to pass in empty GrFlushInfo here since SignalSemaphores()
   // will populate it with semaphores and call GrDirectContext::flush.
   if (shared_context_state_->gr_context()) {
-    skgpu::ganesh::Flush(surface);
+    if (!skgpu::ganesh::Flush(surface).fSuccess) {
+      DLOG(ERROR) << "ClearSharedImage: skgpu::ganesh::Flush failed";
+      return false;
+    }
   } else {
     DCHECK(shared_context_state_->graphite_shared_context());
     DCHECK(shared_context_state_->gpu_main_graphite_recorder());
